sklearn/metrics/pairwise.py

```diff
@@ -1381,6 +1382,10 @@ def pairwise_distances(X, Y=None, metric="euclidean", n_jobs=None, **kwds):
 
     if metric == "precomputed":
         X, _ = check_pairwise_arrays(X, Y, precomputed=True)
+
+        whom = ("`pairwise_distances`. Precomputed distance "
+                " need to have non-negative values.")
+        check_non_negative(X, whom=whom)
```

Comment:
I wonder if there are contexts in which this adds substantial overhead...?

Reply:
Sure, we need to be careful.
The added check corresponds to one full pass over the precomputed distance matrix, so it takes a non-trivial amount of time, but I guess it is not significant compared to most subsequent algorithms.

Note that in this particular case, X is either dense or a CSR sparse matrix, thanks to `check_pairwise_arrays` , so `check_non_negative` does not add any sparse matrix conversion overhead.